File: r/tests/testthat/test-routing.R
test_that("required native igraph routing example works", {
    graph <- obstacle_graph()
    routed <- route_edges(graph, routing = "orthogonal")
    expect_true(igraph::is_igraph(routed))
    expect_equal(igraph::vcount(routed), igraph::vcount(graph))
    expect_equal(igraph::ecount(routed), igraph::ecount(graph))
    routes <- igraph::edge_attr(routed, "adaptagrams_route")
    expect_route_avoids(routes[[1]], c(100, 0), c(40, 40))
})

test_that("orthogonal and polyline routing avoid obstacles", {
    orthogonal <- igraph::edge_attr(
        route_edges(obstacle_graph(), routing = "orthogonal"),
        "adaptagrams_route"
    )[[1]]
    differences <- apply(orthogonal, 2, diff)
    expect_true(all(abs(differences[, 1]) < 1e-9 |
        abs(differences[, 2]) < 1e-9))

    polyline <- igraph::edge_attr(
        route_edges(obstacle_graph(), routing = "polyline"),
        "adaptagrams_route"
    )[[1]]
    expect_true(all(is.finite(polyline)))
    expect_route_avoids(polyline, c(100, 0), c(40, 40))
})

test_that("parallel edges preserve edge identity", {
    graph <- igraph::make_empty_graph(2, directed = TRUE)
    graph <- igraph::add_edges(graph, c(1, 2, 1, 2))
    graph <- igraph::set_vertex_attr(graph, "x", value = c(0, 100))
    graph <- igraph::set_vertex_attr(graph, "y", value = c(0, 0))
    graph <- igraph::set_edge_attr(
        graph,
        "edge_id",
        value = c("first", "second")
    )
    result <- route_edges(graph)
    expect_equal(igraph::ecount(result), 2)
    expect_equal(
        igraph::edge_attr(result, "edge_id"),
        c("first", "second")
    )
    expect_length(igraph::edge_attr(result, "adaptagrams_route"), 2)
})

test_that("self loops produce an edge-specific error", {
    graph <- igraph::make_empty_graph(1, directed = TRUE)
    graph <- igraph::add_edges(graph, c(1, 1))
    graph <- igraph::set_vertex_attr(graph, "x", value = 0)
    graph <- igraph::set_vertex_attr(graph, "y", value = 0)
    expect_error(route_edges(graph), "edge 0 is a self-loop")
    expect_equal(igraph::ecount(graph), 1)
})

test_that("ordinary data frames use the native router", {
    vertices <- data.frame(
        x = c(0, 200, 100),
        y = c(0, 0, 0),
        width = c(80, 80, 40),
        height = c(40, 40, 40)
    )
    edges <- data.frame(source = 1L, target = 2L)
    result <- route_edges(vertices, edges)
    expect_length(result$routes, 1)
    expect_equal(result$edges$adaptagrams_edge_id, "edge-0")
    expect_route_avoids(result$routes[[1]], c(100, 0), c(40, 40))
})

test_that("fixed ports and direction constraints are supported", {
    graph <- obstacle_graph()
    graph <- igraph::set_edge_attr(
        graph,
        "source_port",
        value = list(c(40, 0))
    )
    graph <- igraph::set_edge_attr(
        graph,
        "target_port",
        value = list(c(160, 0))
    )
    graph <- igraph::set_edge_attr(graph, "source_direction", value = "right")
    graph <- igraph::set_edge_attr(graph, "target_direction", value = "left")
    result <- route_edges(
        graph,
        source_port_attribute = "source_port",
        target_port_attribute = "target_port",
        source_direction_attribute = "source_direction",
        target_direction_attribute = "target_direction",
        routing_parameters = list(shape_buffer_distance = 2)
    )
    route <- igraph::edge_attr(result, "adaptagrams_route")[[1]]
    expect_equal(unname(route[1, ]), c(40, 0))
    expect_equal(unname(route[nrow(route), ]), c(160, 0))
})

File: python/tests/test_routing.py
"""Native routing behavior tests."""

import math

import igraph as ig
import pytest
from conftest import assert_route_avoids

import pyadaptagrams as ag


def test_required_routing_example(obstacle_graph: ig.Graph) -> None:
    routed = ag.route_edges(obstacle_graph, routing="orthogonal")
    assert isinstance(routed, ig.Graph)
    assert routed.vcount() == obstacle_graph.vcount()
    assert routed.ecount() == obstacle_graph.ecount()
    assert "adaptagrams_route" in routed.es.attributes()
    assert_route_avoids(routed.es["adaptagrams_route"][0], (100.0, 0.0), (40.0, 40.0))


def test_orthogonal_segments(obstacle_graph: ig.Graph) -> None:
    route = ag.route_edges(obstacle_graph).es["adaptagrams_route"][0]
    for first, second in zip(route, route[1:]):
        assert first[0] == pytest.approx(second[0], abs=1e-9) or first[
            1
        ] == pytest.approx(second[1], abs=1e-9)


def test_polyline_is_finite_and_avoids_obstacle(obstacle_graph: ig.Graph) -> None:
    route = ag.route_edges(obstacle_graph, routing="polyline").es["adaptagrams_route"][
        0
    ]
    assert all(math.isfinite(value) for point in route for value in point)
    assert_route_avoids(route, (100.0, 0.0), (40.0, 40.0))


def test_parallel_edges_remain_addressable() -> None:
    graph = ig.Graph(n=2, edges=[(0, 1), (0, 1)], directed=True)
    graph.vs["x"] = [0.0, 100.0]
    graph.vs["y"] = [0.0, 0.0]
    graph.es["edge_id"] = ["first", "second"]
    result = ag.route_edges(graph)
    assert result.ecount() == 2
    assert result.es["edge_id"] == ["first", "second"]
    assert len(result.es["adaptagrams_route"]) == 2


def test_self_loop_has_specific_error_and_does_not_mutate() -> None:
    graph = ig.Graph(n=1, edges=[(0, 0)], directed=True)
    graph.vs["x"] = [0.0]
    graph.vs["y"] = [0.0]
    before = graph.copy()
    with pytest.raises(ag.UnsupportedGraphError, match="edge 0.*self-loop"):
        ag.route_edges(graph)
    assert graph.get_edgelist() == before.get_edgelist()
    assert graph.vs.attributes() == before.vs.attributes()


def test_fixed_ports_directions_and_parameters(obstacle_graph: ig.Graph) -> None:
    obstacle_graph.es["source_port"] = [(40.0, 0.0)]
    obstacle_graph.es["target_port"] = [(160.0, 0.0)]
    obstacle_graph.es["source_direction"] = ["right"]
    obstacle_graph.es["target_direction"] = ["left"]
    result = ag.route_edges(
        obstacle_graph,
        source_port_attribute="source_port",
        target_port_attribute="target_port",
        source_direction_attribute="source_direction",
        target_direction_attribute="target_direction",
        routing_parameters={"shape_buffer_distance": 2.0, "segment_penalty": 50.0},
    )
    route = result.es["adaptagrams_route"][0]
    assert route[0] == pytest.approx((40.0, 0.0))
    assert route[-1] == pytest.approx((160.0, 0.0))


def test_ordinary_records_use_same_native_router() -> None:
    vertices = [
        {"x": 0.0, "y": 0.0, "width": 80.0, "height": 40.0},
        {"x": 200.0, "y": 0.0, "width": 80.0, "height": 40.0},
        {"x": 100.0, "y": 0.0, "width": 40.0, "height": 40.0},
    ]
    result = ag.route_edges(vertices, [{"source": 0, "target": 1}])
    assert len(result["routes"]) == 1
    assert result["edges"][0]["adaptagrams_edge_id"] == "edge-0"
    assert_route_avoids(result["routes"][0], (100.0, 0.0), (40.0, 40.0))
